### ISSUE TITLE
Rabbit stuck loop issue on network blip

### ISSUE THREAD

**Issue by jbentus** [houseofcat/RabbitMQ.Core#14]
Hi Tristan, how are you ?

On network blips / lost connectivity to Rabbit, `RabbiMQ.Core` using `UseTransientChannels: true` recovers connectivity but it doesn't seem to be disposing the channels used, instead creates new ones, therefore, on every rabbit restart, the lib creates additional channels and consumers. If you keep restarting rabbit or having more network blips, when the number of Channels used reaches the Max number of channels set, the library goes into a loop of continuously creating consumers, connections & channels.

I know this repo is no longer worked on, but at the moment I can't quickly migrate to the new Tesseract solution so I decided to raise an issue here. I've locally tested a change around [here](https://github.com/houseofcat/RabbitMQ.Core/blob/00a810d919669e8e98f252bb7a36c00f9f830ce2/CookedRabbit.Core/Consumer/Consumer.cs#L231-L236), and removed the `&& !(ConsumerSettings.UseTransientChannels ?? true)`. Although it worked it may not be the best solution.

If you're happy receiving a PR based on the above I am happy to make it.

**Comment by jbentus** [comment]
@houseofcat

**Comment by houseofcat** [comment]
Hey Jose, long time no hear from!

I am going to start with an assumption you are all doing something new, either new CI/CD, increased consumer counts, or publishing more than ever before, or the servers have changed. Somethings changed. The assumption is easy to hold in that we have handle handle dozens of outages for years at Fanatics, the last few required no manual intervention. In fact, our apps were the ones identifying issues with our clusters.

I also see no server stats (uptime, churn, connections, or channel counts before and after), or mnesia logs, so I am guessing you guys started digging into the code before troubleshooting your server clusters.

If this is because RabbitMQ is being hosted on transient dockers/K8s that is getting spun up and spun down, then I am going to roll my eyes and with total sincerity tell you that's stupid. Feel free to send my regards to management on that one.

**All RabbitMQ channels that have network issues are dead.** That code you see is to make sure the ChannelHost returns to the ChannelPool (the custom object) and doesn't leaved the ChannelPool drained. When returning it to the "ChannelPool", during a fault, it heals the parent connection and creates a new "channel" inside the ChannelHost.

Since consumers are using transient channels - they are discardable. We use them for Consumers because they would otherwise drain your ChannelPool for publishing. I don't recommend that. It's also why Consumers have more robust code for reconnectivity. There isn't anything to do (or that you can do) with a transient channel on an outage except new one up.

```
On network blips / lost connectivity to Rabbit, RabbiMQ.Core using UseTransientChannels: true recovers connectivity but it doesn't seem to be disposing the channels used
```
As mentioned there is nothing really to do. We can't "dispose" the object in memory as it is re-used/re-assigned to. You can't re-assign to a variable that has been "disposed" it is a runtime exception.

If there is a network outage... the channel is dead. The Consumer only calls SetChannelHostAsync() when going through startup or recovery.

Just a reminder, Channels != Connections, Connections are recoverable, Channels are not.

```
If you keep restarting rabbit or having more network blips, when the number of Channels used reaches the Max number of channels set, the library goes into a loop of continuously creating consumers, connections & channels.
```

Why on earth would you be constantly restarting the RabbitMQ server?
- If this is because RabbitMQ is in Kubernetes cluster, restart the app.
- If this is because you are chaos testing crazy networking stuff, restart the app.
- If this is a prod server constantly blipping in and out of existence, fix the server/clusters. This is a library that helps with transient and temporary server outages, not a substitution for someone nuking a rabbit cluster during planned maintenance. It doesn't perform magic.

Just spit balling what the problem is, your application publishing and consumers have grown but none of the default or original configuration settings have been adjusted.

Max Channels per Connection is a server side setting and it is configurable in the Config. You can edit this:
https://github.com/houseofcat/RabbitMQ.Core/blob/00a810d919669e8e98f252bb7a36c00f9f830ce2/CookedRabbit.Core/Configs/FactoryOptions.cs#L16

It is set conservatively to 100 per connection in anticipation you are load balancing over multiple connections and sockets and need a little space for spike transient usage.

**If you are indeed blocked by "spike" channel growth then configure it correctly.**
 * If you have 5 connections, and 5 channels per Connection, that's 25 total channels.
 * You should set the Max Channel to no lower than 25 so that if they all get recovered on a single Connection in a worst case scenario, it can absorb (slowly of course) how many channels your application is expecting to use.
 
So if you have 50 channels per 10 connections, you have 500 total channels. You would ideally set the the MaxChannel per Connection to 500 to handle a single Connection/Socket doing the full recovery. This does not account for the usage of transient channels (primarily used for Consumers).
 
Lastly, you need to +1 for every consumer you have. So if you have 50 channels per 10 connections (500 total channels) you should add the +25 if you have 25 consumers then x2. This handles the events where a consumer is restarting but the server hasn't quite "let go" of any resources related to the previous transient channel-connection relationship. Bringing the "spike" calculation to about 550 channels.

TL;DR
You should slowly increase `MaxChannelsPerConnection` number till you find that you are handling the outage appropriately.

If I understand what you have done in your code, your code change is just putting transient ChannelHost inside a ChannelPool they didn't come from. Therefore, I reject the idea - its antithetical to the idea of a managed ChannelHosts in the ChannelPool. There aren't designed to be mixed together.

As a heads up the new Consumer which is now 2+ years old at this point is located here:
https://github.com/houseofcat/Tesseract/blob/main/src/HouseofCat.RabbitMQ/Consumer/Consumer.cs

I don't know what code you have, but this may or may not have newer bug fixes and you can see if any of them retro-actively can be applied. I do know the newer Consumer seems healthier and better performing but its in tandem with the cleaner ChannelPool/ConnectionPool.

**Comment by jbentus** [comment]
Hey Tristan,

Hoping you're doing well and really appreciated your reply on this.

I don't know what or if anything has changed about how the Rabbit cluster is being managed, but I can tell you that I've noticed this issue for the first time on 29th October 21 and only recently has happened again since.

In Prod, one of the rabbit nodes in the cluster is restarted (for some reason), and this seems to precede the moment when the loop issue triggers. Looking at the uptime in Rabbit UI, we can see a restart has happened in one node and it matches the timespan of when `RabbitMQ.Core` lib [logs](https://github.com/houseofcat/RabbitMQ.Core/blob/00a810d919669e8e98f252bb7a36c00f9f830ce2/CookedRabbit.Core/Consumer/Consumer.cs#L362-L364) the `shutdown event has occurred (...)` msg.

I can confirm that locally, the issue is reproducible by simulating some 'network blips' through a few rabbit restarts. See screenshot below with some stats while it's happening:

![image](https://user-images.githubusercontent.com/10818547/152641931-1c018b4a-7880-4fa5-9f3a-2ea74b5d77a3.png)

In the picture above, the 100 in the middle is both the number of `MaxChannels` and `Channels`, this is the moment where the library goes into a loop of continuously creating consumers, connections & channels. What seems to have happened building up to this moment is that on every 'network blip' the number of channels and consumers had increased. For that reason, I think increasing the number of max channels decreases the chances of triggering it but does not solve the underlying issue.

> the server hasn't quite "let go" of any resources related to the previous transient channel-connection relationship.

This made me think, it sounded like the client here in this case. What if it's not the server that didn't let go resources but the client because the temporary solution for us when dealing with the outage, is to restart the app and that clears all the resources stuck between client and server.

I have tried your suggested change, I think that helped but it didn't solve it. Throughout testing, I could still see too many channels and consumers being created (but less than before) and it did not 'dispose'/cleared up the channels and consumers previously created between the lib and the server.

**Comment by houseofcat** [comment]
@jbentus 
I have put in a patch to try and help with this kind of behavior.

Basically we are keeping the previous "state" of the consumer in memory. This was done in the new library to remove OLD EventHandlers to try and prevent an eventing leak. This is possibly one cause of the above issue.

Second change was that IF that doesn't solve the "trigger" then put load shedding on the Shutdown/Restart mechanism. That way duplicate calls just fall off. I hope that I was able to capture all of these.

I have also added NET6 support and updated the RabbitMQ client library. It specifically has improvements around connection recovery and EventHandling. It is possible they had some bug fixes in there.

https://github.com/houseofcat/RabbitMQ.Core/compare/release-cr/v3.4.1...cr-v3.5.0
